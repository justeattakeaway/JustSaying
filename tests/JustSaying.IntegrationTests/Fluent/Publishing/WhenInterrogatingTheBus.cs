using JustSaying.TestingFramework;
using Microsoft.Extensions.DependencyInjection;
using Newtonsoft.Json;

namespace JustSaying.IntegrationTests.Fluent.Publishing;

public class WhenInterrogatingTheBus(ITestOutputHelper outputHelper) : IntegrationTestBase(outputHelper)
{
    [AwsFact]
    public async Task Then_The_Interrogation_Result_Should_Be_Returned()
    {
        // Arrange
        var completionSource = new TaskCompletionSource<object>();
        var handler = CreateHandler<SimpleMessage>(completionSource);

        var services = GivenJustSaying()
            .ConfigureJustSaying((builder) =>
            {
                builder.WithLoopbackTopic<SimpleMessage>(UniqueName);
                builder.Publications(p => p.WithTopic<SimpleMessage>());
            })
            .AddSingleton(handler);

        string json = "";
        await WhenAsync(
            services,
            async (publisher, listener, cancellationToken) =>
            {
                await listener.StartAsync(cancellationToken);
                await publisher.StartAsync(cancellationToken);

                var listenerJson = JsonConvert.SerializeObject(listener.Interrogate(), Formatting.Indented);
                var publisherJson = JsonConvert.SerializeObject(publisher.Interrogate(), Formatting.Indented);

                json = string.Join($"{Environment.NewLine}{Environment.NewLine}",
                        listenerJson, publisherJson)
                    .Replace(UniqueName, "integrationTestQueueName", StringComparison.Ordinal);

                completionSource.SetResult(null);
            });

        json.ShouldMatchApproved(opt => opt.SubFolder("Approvals"));
    }
}
