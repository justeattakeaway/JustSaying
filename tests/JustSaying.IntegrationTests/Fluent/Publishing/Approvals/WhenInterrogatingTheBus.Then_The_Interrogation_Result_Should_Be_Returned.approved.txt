{
  "Region": "eu-west-1",
  "Middleware": {
    "Middlewares": [
      {
        "MessageType": "SimpleMessage",
        "QueueName": "integrationTestQueueName",
        "MiddlewareChain": [
          "MessageContextAccessorMiddleware",
          "ErrorHandlerMiddleware",
          "LoggingMiddleware",
          "StopwatchMiddleware",
          "SqsPostProcessorMiddleware",
          "HandlerInvocationMiddleware`1"
        ]
      }
    ]
  },
  "PublishedMessageTypes": [],
  "SubscriptionGroups": {
    "Groups": [
      {
        "Name": "integrationTestQueueName",
        "ConcurrencyLimit": 10,
        "Multiplexer": {
          "ChannelCapacity": 100,
          "ReaderCount": 1
        },
        "ReceiveBuffers": [
          {
            "BufferSize": 10,
            "QueueName": "integrationTestQueueName",
            "Region": "eu-west-1",
            "Prefetch": 10
          }
        ]
      }
    ]
  }
}

{
  "Region": "eu-west-1",
  "Middleware": {
    "Middlewares": []
  },
  "PublishedMessageTypes": [
    "SimpleMessage"
  ],
  "SubscriptionGroups": null
}